### ISSUE TITLE
[Bug] Tests raise `ValueError` when `numpy version` contains non-numerical components

### ISSUE THREAD

**Issue by mgorny** [beartype/beartype#281]
When `numpy` version contains non-numeric components, e.g. it is `1.26.0rc1`, a large number of tests error out while trying to process the version number. An example backtrace is:

```pytb
$ python -m pytest beartype_test/a00_unit/a20_util/hint/a00_pep/test_a90_utilpeptest.py::test_is_hint_pep_supported
========================================================= test session starts =========================================================
platform linux -- Python 3.12.0rc2, pytest-7.4.2, pluggy-1.3.0 -- /usr/lib/python-exec/python3.12/python
cachedir: .pytest_cache
rootdir: /tmp/beartype
configfile: pytest.ini
plugins: forked-1.6.0, xdist-3.3.1, xprocess-0.22.2, asyncio-0.21.1, anyio-4.0.0, pkgcore-0.12.23
asyncio: mode=Mode.STRICT
collected 1 item                                                                                                                      

beartype_test/a00_unit/a20_util/hint/a00_pep/test_a90_utilpeptest.py::test_is_hint_pep_supported FAILED                         [100%]

============================================================== FAILURES ===============================================================
_____________________________________________________ test_is_hint_pep_supported ______________________________________________________

    def test_is_hint_pep_supported() -> None:
        '''
        Test the
        :func:`beartype._util.hint.pep.utilpeptest.is_hint_pep_supported`
        tester.
        '''
    
        # Defer test-specific imports.
        from beartype._util.hint.pep.utilpeptest import is_hint_pep_supported
>       from beartype_test.a00_unit.data.hint.data_hint import (
            NOT_HINTS_UNHASHABLE, NOT_HINTS_PEP)

is_hint_pep_supported = <function is_hint_pep_supported at 0x7ff7e76dd300>

beartype_test/a00_unit/a20_util/hint/a00_pep/test_a90_utilpeptest.py:203: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
beartype_test/a00_unit/data/hint/data_hint.py:23: in <module>
    from beartype_test.a00_unit.data.hint.pep.data_pep import (
        AnyType    = <class 'object'>
        HINTS_NONPEP_META = (HintNonpepMetadata(
    hint=<class 'beartype_test.a00_unit.data.hint.nonpep.proposal._data_nonpep484.NamedTupleType'...exes=('\\bcomplex\\b', '\\btype hint\\b'),
    exception_str_not_match_regexes=('\\n', '\\*', '"complex"'),
)),
), ...)
        NoneType   = <class 'NoneType'>
        NoneTypeOr = {}
        __builtins__ = <builtins>
        __cached__ = '/tmp/beartype/beartype_test/a00_unit/data/hint/__pycache__/data_hint.cpython-312.pyc'
        __doc__    = '\n**Beartype type hints data-driven testing submodule.**\n\nThis submodule predefines low-level global constants exer... test submodules -- including PEP-compliant\ntype hints, PEP-noncompliant type hint, and objects satisfying neither.\n'
        __file__   = '/tmp/beartype/beartype_test/a00_unit/data/hint/data_hint.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7ff7e745c740>
        __name__   = 'beartype_test.a00_unit.data.hint.data_hint'
        __package__ = 'beartype_test.a00_unit.data.hint'
        __spec__   = ModuleSpec(name='beartype_test.a00_unit.data.hint.data_hint', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7ff7e745c740>, origin='/tmp/beartype/beartype_test/a00_unit/data/hint/data_hint.py')
beartype_test/a00_unit/data/hint/pep/data_pep.py:203: in <module>
    _init()
        HINTS_PEP_HASHABLE = None
        HINTS_PEP_IGNORABLE_DEEP = set()
        HINTS_PEP_IGNORABLE_SHALLOW = {<class 'object'>}
        HINTS_PEP_META = []
        __builtins__ = <builtins>
        __cached__ = '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/__pycache__/data_pep.cpython-312.pyc'
        __doc__    = '\n**Beartype PEP-compliant type hints test data.**\n\nThis submodule predefines low-level global constants whose values are\nPEP-compliant type hints, exercising known edge cases on behalf of higher-level\nunit test submodules.\n'
        __file__   = '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/data_pep.py'
        __loader__ = <_frozen_importlib_external.SourceFileLoader object at 0x7ff7e7495730>
        __name__   = 'beartype_test.a00_unit.data.hint.pep.data_pep'
        __package__ = 'beartype_test.a00_unit.data.hint.pep'
        __spec__   = ModuleSpec(name='beartype_test.a00_unit.data.hint.pep.data_pep', loader=<_frozen_importlib_external.SourceFileLoader object at 0x7ff7e7495730>, origin='/tmp/beartype/beartype_test/a00_unit/data/hint/pep/data_pep.py')
        _init      = <function _init at 0x7ff7e7483a60>
beartype_test/a00_unit/data/hint/pep/data_pep.py:175: in _init
    data_hint_pep_submodule.add_data(CURRENT_SUBMODULE)
        CURRENT_SUBMODULE = <module 'beartype_test.a00_unit.data.hint.pep.data_pep' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/data_pep.py'>
        DATA_HINT_PEP_SUBMODULES = (<module 'beartype_test.a00_unit.data.hint.pep.mod._data_hintmodnumpy' from '/tmp/beartype/beartype_test/a00_unit/data...nt.pep.proposal._data_pep589' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep589.py'>, ...)
        HintPepMetadata = <class 'beartype_test.a00_unit.data.hint.util.data_hintmetacls.HintPepMetadata'>
        _data_hintmodnumpy = <module 'beartype_test.a00_unit.data.hint.pep.mod._data_hintmodnumpy' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/mod/_data_hintmodnumpy.py'>
        _data_pep544 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep544' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep544.py'>
        _data_pep585 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep585' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep585.py'>
        _data_pep586 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep586' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep586.py'>
        _data_pep589 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep589' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep589.py'>
        _data_pep593 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep593' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep593.py'>
        _data_pep604 = <module 'beartype_test.a00_unit.data.hint.pep.proposal._data_pep604' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/_data_pep604.py'>
        data_hint_pep_submodule = <module 'beartype_test.a00_unit.data.hint.pep.mod._data_hintmodnumpy' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/mod/_data_hintmodnumpy.py'>
        data_pep484 = <module 'beartype_test.a00_unit.data.hint.pep.proposal.data_pep484' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/proposal/data_pep484.py'>
        is_object_hashable = <function is_object_hashable at 0x7ff7e767e660>
        sys        = <module 'sys' (built-in)>
beartype_test/a00_unit/data/hint/pep/mod/_data_hintmodnumpy.py:41: in add_data
    if not is_package_numpy_typing_ndarray_deep():
        data_module = <module 'beartype_test.a00_unit.data.hint.pep.data_pep' from '/tmp/beartype/beartype_test/a00_unit/data/hint/pep/data_pep.py'>
beartype/_util/cache/utilcachecall.py:278: in _callable_cached
    raise exception
        args       = ()
        args_flat  = ()
        args_flat_to_exception = {(): ValueError("invalid literal for int() with base 10: '0rc1'")}
        args_flat_to_exception_get = <built-in method get of dict object at 0x7ff7e74c2340>
        args_flat_to_return_value = {}
        args_flat_to_return_value_get = <built-in method get of dict object at 0x7ff7e74c2180>
        func       = <function is_package_numpy_typing_ndarray_deep at 0x7ff7e74d04a0>
        return_value = <beartype._util.utilobject.Iota object at 0x7ff7e780e060>
beartype/_util/cache/utilcachecall.py:270: in _callable_cached
    return_value = args_flat_to_return_value[args_flat] = func(
        args       = ()
        args_flat  = ()
        args_flat_to_exception = {(): ValueError("invalid literal for int() with base 10: '0rc1'")}
        args_flat_to_exception_get = <built-in method get of dict object at 0x7ff7e74c2340>
        args_flat_to_return_value = {}
        args_flat_to_return_value_get = <built-in method get of dict object at 0x7ff7e74c2180>
        func       = <function is_package_numpy_typing_ndarray_deep at 0x7ff7e74d04a0>
        return_value = <beartype._util.utilobject.Iota object at 0x7ff7e780e060>
beartype_test/_util/mod/pytmodtest.py:116: in is_package_numpy_typing_ndarray_deep
    is_package_numpy() and
beartype/_util/cache/utilcachecall.py:278: in _callable_cached
    raise exception
        args       = ()
        args_flat  = ()
        args_flat_to_exception = {(): ValueError("invalid literal for int() with base 10: '0rc1'")}
        args_flat_to_exception_get = <built-in method get of dict object at 0x7ff7e74c1bc0>
        args_flat_to_return_value = {}
        args_flat_to_return_value_get = <built-in method get of dict object at 0x7ff7e74c0f40>
        func       = <function is_package_numpy at 0x7ff7e74d0360>
        return_value = <beartype._util.utilobject.Iota object at 0x7ff7e780e060>
beartype/_util/cache/utilcachecall.py:270: in _callable_cached
    return_value = args_flat_to_return_value[args_flat] = func(
        args       = ()
        args_flat  = ()
        args_flat_to_exception = {(): ValueError("invalid literal for int() with base 10: '0rc1'")}
        args_flat_to_exception_get = <built-in method get of dict object at 0x7ff7e74c1bc0>
        args_flat_to_return_value = {}
        args_flat_to_return_value_get = <built-in method get of dict object at 0x7ff7e74c0f40>
        func       = <function is_package_numpy at 0x7ff7e74d0360>
        return_value = <beartype._util.utilobject.Iota object at 0x7ff7e780e060>
beartype_test/_util/mod/pytmodtest.py:93: in is_package_numpy
    return is_module_version_at_least(
        _LIB_RUNTIME_OPTIONAL_VERSION_MINIMUM_NUMPY = '1.21.0'
        is_module_version_at_least = <function is_module_version_at_least at 0x7ff7e7700900>
beartype/_util/mod/utilmodtest.py:215: in is_module_version_at_least
    version_actual_parts  = _convert_version_str_to_tuple(version_actual)
        get_module_version = <function version at 0x7ff7e8397c40>
        module_name = 'numpy'
        version_actual = '1.26.0rc1'
        version_minimum = '1.21.0'
beartype/meta.py:117: in _convert_version_str_to_tuple
    return tuple(int(version_part) for version_part in version_str.split('.'))
        version_str = '1.26.0rc1'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <list_iterator object at 0x7ff7e61ab6d0>

>   return tuple(int(version_part) for version_part in version_str.split('.'))
E   ValueError: invalid literal for int() with base 10: '0rc1'

.0         = <list_iterator object at 0x7ff7e61ab6d0>
version_part = '0rc1'

beartype/meta.py:117: ValueError
======================================================= short test summary info =======================================================
FAILED beartype_test/a00_unit/a20_util/hint/a00_pep/test_a90_utilpeptest.py::test_is_hint_pep_supported - ValueError: invalid literal for int() with base 10: '0rc1'
========================================================== 1 failed in 0.28s ==========================================================
```

**Comment by leycec** [comment]
**Gah!** You're absolutely right; that's shockingly bad. I'll get this immediately patched for you and the intrepid Gentoo crew along with an unrelated resolution for #283, which is even worse. @beartype 0.16.2 is shortly to follow.

**Relatedly:** @beartype 0.16.1 just dropped and resolves a dramatic conflict with PyTorch at #282. Yes, it is that bad. It's even worse than that bad. @beartype 0.16.0 should be blacklisted as known-to-be-really-bad. This is why we can't have good things, @mgorny.

**Comment by leycec** [comment]
**Resolved by 771b9d79b.** @beartype's internal routines for converting and comparing version strings are now *considerably* more robust and permissive against horrible edge cases (like NumPy release candidates). In theory, this has been extensively unit tested; @beartype's test suite should now *always* do the right thing regardless of whether the currently installed version of anything is a stable, alpha, beta, or release candidate release. In practice, I promise nothing and deliver even less.

Thanks as always for being so awesome, @mgorny. You make Gentoo the greatest place to be. :+1:
